Merge upstream

richard-cox opened the pull request.

richard-cox commented on src/app/features/applications/application/application-tabs-base/tabs/build-tab/build-tab.component.html:
```diff
@@ -118,3 +118,9 @@ <h3 mat-subheader>Cloud Foundary</h3>
   </div>
+  <div class="summary--routes-card">
+    <app-details-card title="Routes" [busy]="cardTwoFetching$ | async">
+     <app-routes></app-routes>
```
We should move this outside of the card

irfanhabib replied:
done

richard-cox commented on src/app/features/applications/application/application-tabs-base/tabs/build-tab/build-tab.component.ts:
```diff
@@ -1,3 +1,4 @@
+import { NewRoute } from '../../../../../../store/actions/route.actions';
```
To nuke

irfanhabib replied:
done

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+<div>
```
This should have something like the below (to give it the blue header bar)
```
<app-page-header>
  Create Route
</app-page-header>
```

irfanhabib replied:
Done

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+  <app-details-card title="Add Route">
+    <form class="add-route--form" [formGroup]="addRoute" >
+      Domain
+      <mat-form-field>
```
We should have a max width for these. If there's not something global then we should set up one

irfanhabib replied:
I've picked `1920px` as max-width for my form.

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+      </mat-form-field>
+      Host
+      <mat-form-field>
+        <input matInput formControlName="host" placeholder="Host">
```
Are any of these required? If so need the `required` attribute

irfanhabib replied:
Only required is the `domain` field. All others can be optional.

irfanhabib replied:
I've split the form into two different forms. The forms are toggled by a checkbox. Its easier to split it so that we can add `required` fields to the appropriate route type.

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+<div>
+  <mat-progress-bar *ngIf="submitted" [mode]="'indeterminate'">
+  </mat-progress-bar>
+  <app-details-card title="Add Route">
```
Not sure this is the right  element. The collapse part isn't needed

irfanhabib replied:
Switched to a `mat-card`

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+  </mat-progress-bar>
+  <app-details-card title="Add Route">
+    <form class="add-route--form" [formGroup]="addRoute" >
+      Domain
```
In this new world we shouldn't need to label selects or inputs (the placeholder will float above any value added)

irfanhabib replied:
done

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+        <input matInput formControlName="port" placeholder="Port">
+      </mat-form-field>
+      <div class="add-route--form-buttons">
+        <button mat-button name="submit" color="primary" (click)="onSubmit()">
```
The submit button stays valid throughout, should tie into any red/invalid state on fields

irfanhabib replied:
fixed

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.ts:
```diff
@@ -0,0 +1,127 @@
+    this.space$ = this.store.select(selectEntity('application', this.appGuid))
+    .pipe(
+      tap(p => {
+        if (p) {
```
instead of `if(p)` you can place a filter before the tap. For example `filter(p => !!p`)

irfanhabib replied:
done

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.ts:
```diff
@@ -0,0 +1,127 @@
+          this.domains$ = this.store.select(selectNestedEntity('space', this.spaceGuid, ['entity', 'domains']))
+          .pipe(
+            tap(d => {
+              if (d) {
```
Same here too (filter)

irfanhabib replied:
done

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.ts:
```diff
@@ -0,0 +1,127 @@
+      tap(p => {
+        if (p) {
+          this.spaceGuid = p.entity.space_guid;
+          this.domains$ = this.store.select(selectNestedEntity('space', this.spaceGuid, ['entity', 'domains']))
```
Not sure if there's any point where this tap is executed again, but we should be safe and unsub as in destroy

irfanhabib replied:
Good catch! Fixed

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,34 @@
+    <form class="add-route--form" [formGroup]="addRoute" >
+      Domain
+      <mat-form-field>
+        <mat-select  formControlName="domain" required>
```
For drop downs we should select the first item. This isn't a pattern we've used elsewhere but we should have done.

irfanhabib replied:
done

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,48 @@
+  <app-details-card title="Add Route">
+    <div class="add-route--form">
+      <div class="add-route--form-toggle">
+          Create TCP Route
```
We probably need to tweak this. Something like a check box which disabled/enables the specific fields

irfanhabib replied:
done

richard-cox commented:
I jumped the gun with the review. Some of the form comments aren't valid anymore.

richard-cox commented on src/app/features/applications/routes/add-routes/add-routes.component.html:
```diff
@@ -0,0 +1,77 @@
+          </div>
+          <div>
+            <mat-form-field>
+              <input matInput formControlName="port" placeholder="Port" required>
```
Could add a simple number reg ex, something like...
```
            <mat-form-field>
              <input matInput formControlName="port" placeholder="Port" required>
              <mat-error *ngIf="addTCPRoute.controls['port'].errors?.pattern">Invalid Port</mat-error>
            </mat-form-field>
```
and
```
port: new FormControl('', [Validators.required, Validators.pattern('[0-9]*')])
```
If there's other places where we can apply some basic filtering we should do

richard-cox commented on src/app/shared/components/table/custom-cells/table-cell-tcproute/table-cell-tcproute.component.html:
```diff
@@ -0,0 +1,8 @@
+    <mat-icon>check</mat-icon>
+  </div>
+  <div *ngIf="!isRouteTCP">
+    <mat-icon>close</mat-icon>
```
We should swap this out with something else, the cross might be confused with a delete. In other places we've just used a dash `-` for missing entries

irfanhabib commented:
@richard-cox I've updated the PR.
The issue displaying with `undefined` showing in the table after an update of the table is due to the update requests not including the initialParams. I've captured the issue here; https://github.com/SUSE/stratos/issues/114

richard-cox committed:
Fix list loading/refresh indicator
- Specific refresh indicator based on click of refresh button
- New e2e test for non-local space routes list
- Prep for e2e loading/refresh indicator tests

richard-cox committed:
Merge remote-tracking branch 'origin/v2-master' into docker-instructions-tweaks

richard-cox committed:
Tweak cf push instructions
- group together cf push from source information
- add 1 line cf push docker image instructions

richard-cox committed:
Fix cc issue

richard-cox committed:
Fix route links, Improve map route e2e

richard-cox committed:
Fix marketplace/build pack card icon

richard-cox committed:
Fix for initial load indicator

richard-cox committed:
e2e test fixes

nwmac committed:
Merge pull request #3314 from cloudfoundry-incubator/docker-instructions-tweaks

Tweak cf push instructions

nwmac committed:
Helm Chart Upgrade fix (#3316)

* Helm Chart Upgrade fix

* Fix placement of db selector

* Added test script

richard-cox committed:
Fix e2e test

richard-cox committed:
Fix double requests when single endpoint connected

nwmac committed:
Fix issues with automated helm test (#3317)

* Ensure we support spaces in paths

* Better fix for spaces in file paths

* Debugging

* Space in names fix

* Fix path issues with spaces

* Fixed temp dir creation bug

* Add missing $ to var in copy command

nwmac committed:
Don't stop and start PCF Dev each time before/after tests (#3318)

nwmac committed:
Reduce weight of nav logo (#3321)

richard-cox committed:
Fix null exception after creating a space in an new org
- We previously nuked the org's space collection when creating or deleting an org
- Validation was correctly then going to refetch this list, however waitForEntity is broken (see #2327) and emits an invalid action
- Later code takes invalid action an tries to access space
- Fix, aside from #2327, is yodate org space collection after space creation instead of nuking and waiting for validation to refetch
Other fixes
- Found sub leak in edit org stepper
- endpoint service allApps is subbed to in many places, ensure we share

richard-cox committed:
Fix search by space name in manage users space roles table
- fixes #3329

richard-cox committed:
Remove full stops from empty list messages
- fixes #3340

nwmac committed:
Merge pull request #3315 from cloudfoundry-incubator/route-changes

Route & Card icon fixes

KlapTrap committed:
Merge pull request #3354 from cloudfoundry-incubator/small-fixes

Fix manager users space table search & improve 'too many x' text

Commit:
Tidy up org spce reducer logic

KlapTrap committed:
Merge pull request #3351 from cloudfoundry-incubator/blank-org-details

Fix null exception after creating a space in an new org

nwmac committed:
Fix filter panel size on services and marketplace pages (#3358)

KlapTrap committed:
Manually make the date selector date a moment object (#3352)

nwmac committed:
Ensure the SCM type is included in store keys (#3353)

* Ensure the SCM type is included in store keys

* Fix multi scm tab on env var update and 'Redeploy to update' message now correctly shown

* Fix multiple tabs issue

nwmac committed:
Fix display of percent tooltip on metrics chart (#3361)

nwmac committed:
Fix app state to allow start action for offline while updating (#3350)

* Fix app state to allow start action for offline while updating

* Remove fdescribe

Commit:
Revert loading indicator fixes

nwmac committed:
About Page Title customization support (#3356)

* Add optional title text to title component

* Support custom about page info component

* Undo typo

nwmac committed:
Merge pull request #3313 from cloudfoundry-incubator/fix-list-loading-indicators

Fixes double requests when single cf connected for lists with cf filter

nwmac committed:
Fix CF-Level Metrics permissions check (#3348)

* Fix CF-Level Metrics permissions check

* Remove debug log message

* Fix non-cf admin metrics
- Cell metrics used incorrect endpoint (regression)
- This whitelists certain metrics that a non-cf admin can use
- Non-cf admins can see all cell details... except apps in cells
- Also fixed issue where cell list used old action

richard-cox committed:
Fix fetching of stats in app wall for some cases when list is maxed (#3360)

- Allow time for the list to settle
- fixes #3341

KlapTrap committed:
Ensure the loading indicator and the refresh button animated at the correct time. (#3363)

* Fix list component refresh vs. loading logic

* Ensure we don't needlessly modify the pagination client filter object

* Ensure the refresh button animates on first click on non-local lists

* Tidy up

* Merge fixes

nwmac committed:
Change log for 2.3.0 (#3364)

* Change log for 2.3.0

* Fixed `JSON Schemas ` line

* Fixed JSON scheme line link

nwmac committed:
Weekly updaye 18 January (#3365)

richard-cox committed:
Merge remote-tracking branch 'origin/v2-master' into merge-upstream

richard-cox committed:
Merge remote-tracking branch 'upstream/v2-master' into merge-upstream

nwmac committed:
Fix for bug causing intermittment E2E Test failures on org tests (#3366)

richard-cox committed:
Merge remote-tracking branch 'upstream/v2-master' into merge-upstream

richard-cox committed:
Fix bleed of mat-card style

nwmac reviewed (approved):
LGTM
